File: src/stack/1.cpp
/**

PUSh_STACK (stack, top, max, item)
  1. if top = max then
      print 'stack is full, Overflow' and return
  2. top:= top + 1
  3. stack(top):= item
  4. exit

 */

#include <iostream>
using namespace std;

int PUSH_stack(int stack[], int &top, int max, int item) {
  if (top == max - 1) { // Check if the stack is full
    cout << "Stack is full, Overflow!" << endl;
    return 0;
  }

  top++;             // Increment the top index
  stack[top] = item; // Push the item onto the stack
  return 1;          // Return 1 to indicate success
}

int main() {
  cout << "Push Operation in Stack\n";
  int max = 5;
  int top = -1; // Initialize top to -1 to indicate an empty stack
  int stack[max];
  int item;

  while (true) {
    cout << "Enter an item to push (or any non-integer to exit): ";
    cin >> item;

    if (cin.fail()) {
      cout << "Exiting program." << endl;
      break; // Exit the loop if the input is not an integer
    }

    int res = PUSH_stack(stack, top, max, item);
    if (res == 0)
      break;
  }

  return 0;
}

File: src/stack/1.ts
import prompt from "prompt-sync";

class Stack {
  private stack: number[];
  private top: number;
  private max: number;

  constructor(max: number) {
    this.max = max;
    this.stack = new Array(max);
    this.top = -1;
  }

  push(item: number): void {
    if (this.top === this.max - 1) {
      console.log("Stack is full, Overflow!");
      return;
    }
    this.top++;
    this.stack[this.top] = item;
  }
  pop(): number {
    if (this.top === -1) {
      console.log("Stack is Empty, Underflow!");
      return -1;
    }
    const item = this.stack[this.top];
    this.top--;
    return item;
  }
}

const main = () => {
  const max = 5;
  const stack = new Stack(max);
  const promptSync = prompt();

  while (true) {
    console.log("stack Operations:");
    console.log("1. Push");
    console.log("2. Pop");
    console.log("3. Quit");
    const choice = parseInt(promptSync("Enter Your Choice: "));

    switch (choice) {
      case 1: {
        const item = parseInt(promptSync("Enter an item to push: "));
        stack.push(item);
        break;
      }
      case 2: {
        const poppedItem = stack.pop();
        if (poppedItem !== -1) {
          console.log(`Popped Item: ${poppedItem}`);
        }
        break;
      }
      case 3: {
        console.log("Exiting Program.");
        return;
      }
      default:
        console.log("Invalid Choice. Try Again.");
        break;
    }
  }
};

main();
